File: TEK3/DOTNET/DOT_cardGames_2017/Server/GameRules/Deck.cs
using System;
using System.Collections.Generic;

namespace Coinche.GameRules
{
    /// <summary>
    /// class defining the deck
    /// </summary>
    public class Deck
    {
        private List<Card> Cards;
        private int CurrentIdx;
        
        /// <summary>
        /// Basic constructor that creates the deck using the Card class
        /// </summary>
        public Deck()
        {
            Cards = new List<Card>();

            foreach (CardColor c in Enum.GetValues(typeof(CardColor)))
            {
                foreach (CardValue v in Enum.GetValues(typeof(CardValue)))
                {
                    Cards.Add(new Card(c, v));
                }
            }
        }

        /// <summary>
        /// Reset the deck state before dealing
        /// </summary>
        public void Reset()
        {
            CurrentIdx = 0;
        }

        /// <summary>
        /// Display the cards in the whole deck
        /// </summary>
        public void Display()
        {
            int i = 1;

            foreach (Card c in Cards)
            {
                Console.WriteLine("Card n°" + i++ + " : " + c.Value + " of " + c.Color);
            }
        }

        /// <summary>
        /// Shuffle the deck in a random order
        /// </summary>
        public void Shuffle()
        {
            List<Card> randomList = new List<Card>();
            Random r = new Random();
            int randomIndex = 0;

            while (Cards.Count > 0)
            {
                randomIndex = r.Next(0, Cards.Count);
                randomList.Add(Cards[randomIndex]);
                Cards.RemoveAt(randomIndex);
            }
            Cards = randomList;
        }

        /// <summary>
        /// Deal 8 cards to a specified player
        /// </summary>
        /// <param name="p"></param>
        public void Deal(Player p)
        {
            for (int i = 0; i < 8; i++)
            {
                p.Hand.Add(Cards[CurrentIdx++]);
            }
        }
    }
}

File: TEK3/Java/Java_jcoinche_2017/modules/rules/src/main/java/com/jcoinche/rules/Deck.java
package com.jcoinche.rules;

import java.util.ArrayList;
import java.util.List;
import static java.util.Collections.shuffle;

/**
 * This class represents the deck we use for the coinche
 *
 * A deck is composed of 32 cards (8 per color from the Seven to the Ace)
 *
 * @author uberti_l
 * @author grella_c
 * @see Card
 * @version 1.0
 */
public class Deck {
    /**
     * the list of the deck's cards
     */
    public List<Card>  deck;
    /**
     * the index from where we deal the cards
     */
    public int         idx;

    /**
     * Basic constructor that initialize the deck.
     */
    Deck() {
        deck = new ArrayList<>();
        idx = 0;

        int num = 0;
        for (Card.CardName name : Card.CardName.values()) {
            for (Card.CardType type : Card.CardType.values()) {
                deck.add(new Card(name, type, num));
            }
            num++;
        }
    }

    /**
     * shuffle the deck before dealing it.
     */
    public void shuffleDeck() {
        shuffle(deck);
        idx = 0;
    }

    /**
     * deal 8 cards to the player p
     */
    public void deal(Player p) {
        p.hand.clear();
        for (int i = 0; i < 8; i++) {
            p.hand.add(deck.get(idx++));
        }
    }
}
